src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/induction/FutureWorkInterestsEntityMapper.kt

```diff
@@ -62,33 +61,38 @@ abstract class FutureWorkInterestsEntityMapper {
     dto: UpdateFutureWorkInterestsDto?,
   )
 
+  fun updateInterests(entity: FutureWorkInterestsEntity, dto: UpdateFutureWorkInterestsDto): List<WorkInterestEntity> {
+    val existingInterests = entity.interests!!
+    val updatedInterests = dto.interests
+
+    entityListManager.updateExisting(existingInterests, updatedInterests, workInterestEntityMapper)
+    entityListManager.addNew(entity, existingInterests, updatedInterests, workInterestEntityMapper)
+    entityListManager.deleteRemoved(existingInterests, updatedInterests)
+
+    return existingInterests
+  }
+
+  @BeanMapping(qualifiedByName = ["addNewInterestsDuringUpdate"])
   @ExcludeJpaManagedFieldsIncludingDisplayNameFields
   @GenerateNewReference
   @Mapping(target = "createdAtPrison", source = "prisonId")
   @Mapping(target = "updatedAtPrison", source = "prisonId")
   @Mapping(target = "interests", ignore = true)
   abstract fun fromUpdateDtoToEntity(dto: UpdateFutureWorkInterestsDto?): FutureWorkInterestsEntity
 
+  @Named("addNewInterestsDuringUpdate")
   @AfterMapping
-  fun addInterests(dto: UpdateFutureWorkInterestsDto, @MappingTarget entity: FutureWorkInterestsEntity) {
-    val existingTypes = entity.interests?.map { it.key() }
-    dto.interests.forEach {
-      // ensure we only add new interests
```

Comment:
A colleague of mine (😉 ) pointed out that you can control which client method calls an `@AfterMapping` method. Upon reading the MapStruct docs (🙄 ), I've discovered that using `@BeanMapping(qualifiedByName = ["addNewInterestsDuringUpdate"])` on the originating method, means that `addNewInterestsDuringUpdate` is only called from `fromUpdateDtoToEntity()` and not from `updateEntityFromDto()` as well. This means we no longer need to check if a child entity already exists in the collection before adding it.